Incremental logging time for kube-proxy & kubelet

claudiubelu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

kube-proxy can generate a lot of logs just in standby, making it difficult to analyze them, especially with log rotation, plus, over time, they can start to  occupy quite a sizable amount of disk space.

This PR adds incremental back-off for identical log messages, increasing the logging threshold time by 10 seconds each time the message is logged, up to a maximum of 5 minutes. This means that after a while, frequent log messages are only logged once evewry 5 minutes.

If a message has not been logged for longer than 10 minutes, its backoff resets.

NOTE: the logging implementation should probably be added to klog.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Partially Fixes #78023

**Special notes for your reviewer**:

The amount of logged messages reduced significantly in the span of 1 hour, and the reduction would be more significant over a longer period of time:

old: https://paste.ubuntu.com/p/5h62yPv4NC/
new: https://paste.ubuntu.com/p/TYZw7yxHtr/

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

claudiubelu committed:
Incremental logging time for kube-proxy

kube-proxy can generate a lot of logs just in standby, making it difficult to
analyze them, especially with log rotation, plus, over time, they can start to
occupy quite a sizable amount of disk space.

This commit adds incremental back-off for identical log messages, increasing the
logging threshold time by 10 seconds each time the message is logged, up to a
maximum of 5 minutes. This means that after a while, frequent log messages are
only logged once every 5 minutes.

If a message has not been logged for longer than 10 minutes, its backoff resets.

PatrickLang commented:
cc @ksubrmnn can you review?

vllry commented:
Hi, thanks for the PR!

Do you have a log size comparison between original and updated? EG running in similar circumstances for 20 minutes.

Secondly, I'm a little worried about the size of the time window - whilst that would cut down on a lot of logs, I can see it being confusing for operators.

"kube-proxy generates a lot of logs, even when idle" would be a good root issue to file. I'm not sure if this is the right solution, but I appreciate the problem.

claudiubelu committed:
Incremental logging time for kubelet

kubelet can also generate a lot of logs just in standby, which can
make it hard to analize them.

This will add incremental back-off for some of the identical log
messages, increasing the logging threshold time by 10 seconds each time
the message is logged, up to a maximum of 5 minutes.

If a message has not been logged for longer than 10 minutes, its backoff
resets.

bclau commented:
> 
> 
> Hi, thanks for the PR!
> 
> Do you have a log size comparison between original and updated? EG running in similar circumstances for 20 minutes.

I already had included that in the PR message. Running it for 1 hour: yielded 947 lines of log without these changes, and 645 lines with these changes. Keep in mind that maximum back-off is not achieved in 1 hour.

Running kube-proxy for longer, it can be seen that:

- with these changes, in 24 hours: 336KB logs
- without these changes, in 22 hours: 984KB logs.

So, these reduces the amount of unnecessary logging by ~70%.

> 
> Secondly, I'm a little worried about the size of the time window - whilst that would cut down on a lot of logs, I can see it being confusing for operators.

You mean the increasing time window between consecutive log messages? Why would it be confusing?

> 
> "kube-proxy generates a lot of logs, even when idle" would be a good root issue to file. I'm not sure if this is the right solution, but I appreciate the problem.

I already had an issue opened, and there were a couple of ideas there, and this implementation was one of them: https://github.com/kubernetes/kubernetes/issues/78023

ksubrmnn commented:
/approve for proxy/winkernel changes

bclau commented:
I've added a commit addressing kubelet as well. After a day, the log sizes after 24 hours are as follows:

- old: 41MB
- new: 28MB

Ofc, this did not include the original issue's flooding log messages (the same log message was repeated every 2 seconds).

vllry commented:
> > Secondly, I'm a little worried about the size of the time window - whilst that would cut down on a lot of logs, I can see it being confusing for operators.
> 
> You mean the increasing time window between consecutive log messages? Why would it be confusing?

Skipping logs creates an inaccurate log stream - a user attempting to debug will no longer see the same sequence of events, as some have been censored. Personally I think that 10 minutes is a large window to do that in, but I'll punt that decision.

/assign @thockin

bclau commented:
> 
> 
> > > Secondly, I'm a little worried about the size of the time window - whilst that would cut down on a lot of logs, I can see it being confusing for operators.
> > 
> > 
> > You mean the increasing time window between consecutive log messages? Why would it be confusing?
> 
> Skipping logs creates an inaccurate log stream - a user attempting to debug will no longer see the same sequence of events, as some have been censored. Personally I think that 10 minutes is a large window to do that in, but I'll punt that decision.
> 
> /assign @thockin

It is 5 minutes at most, and it doesn't start with that value. It starts with a 10 second window in which it won't log the same message again, then after 10 seconds, it will. Then a 20 second window, etc. The 5 minute cap will not be reached in 1 hour. Meaning that if one would want to debug by reading the logs, that person probably won't even notice the log message are being backed off initially.

Additionally, I didn't add this behaviour to anything related to pods, pod probes, services, etc (with the exception of the orphaned pod error, which was being logged every 2 seconds, flooding the logs and rendering them unusable). I've added this to mostly system bits that are being logged on a loop that are unlikely to change over time (and if they do, they'll probably log other things instead).

thockin commented:
Log scrubbing is the sort of thing that has to be done frequently.  An
interesting bit of release process might be to capture logs during a series
of controlled operations and then require a signoff from the SIG.

On Mon, Jun 24, 2019 at 1:18 PM Claudiu Belu <notifications@github.com>
wrote:

> Secondly, I'm a little worried about the size of the time window - whilst
> that would cut down on a lot of logs, I can see it being confusing for
> operators.
>
> You mean the increasing time window between consecutive log messages? Why
> would it be confusing?
>
> Skipping logs creates an inaccurate log stream - a user attempting to
> debug will no longer see the same sequence of events, as some have been
> censored. Personally I think that 10 minutes is a large window to do that
> in, but I'll punt that decision.
>
> /assign @thockin <https://github.com/thockin>
>
> It is 5 minutes at most, and it doesn't start with that value. It starts
> with a 10 second window in which it won't log the same message again, then
> after 10 seconds, it will. Then a 20 second window, etc. The 5 minute cap
> will not be reached in 1 hour. Meaning that if one would want to debug by
> reading the logs, that person probably won't even notice the log message
> are being backed off initially.
>
> Additionally, I didn't add this behaviour to anything related to pods, pod
> probes, services, etc (with the exception of the orphaned pod error, which
> was being logged every 2 seconds, flooding the logs and rendering them
> unusable). I've added this to mostly system bits that are being logged on a
> loop that are unlikely to change over time (and if they do, they'll
> probably log other things instead).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/79139?email_source=notifications&email_token=ABKWAVFLVNAOOAQIDTACHKTP4ET2PA5CNFSM4HZAOVR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYOC4CA#issuecomment-505163272>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWAVADSQQBE2G4N4ABAX3P4ET2PANCNFSM4HZAOVRQ>
> .
>

thockin commented on pkg/kubelet/config/file.go:
```diff
@@ -95,7 +96,7 @@ func (s *sourceFile) run() {
 		// Read path immediately to speed up startup.
 		if err := s.listConfig(); err != nil {
-			klog.Errorf("Unable to read config path %q: %v", s.path, err)
+			freqlog.ErrorInfreqf("Unable to read config path %q: %v", s.path, err)
```
Doesn't this get logged exactly once?

thockin commented on pkg/kubelet/kubelet_volumes.go:
```diff
@@ -151,7 +152,7 @@ func (kl *Kubelet) cleanupOrphanedPodDirs(pods []*v1.Pod, runningPods []*kubecon
 	logSpew := func(errs []error) {
 		if len(errs) > 0 {
-			klog.Errorf("%v : There were a total of %v errors similar to this. Turn up verbosity to see them.", errs[0], len(errs))
+			freqlog.ErrorInfreqf("%v : There were a total of %v errors similar to this. Turn up verbosity to see them.", errs[0], len(errs))
```
You want to rate-limit the log that says logs were rate-limited?  How meta...

thockin commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -736,7 +737,7 @@ func VolumeLimits(volumePluginListFunc func() []volume.VolumePluginWithAttachLim
 			attachLimits, err := volumePlugin.GetVolumeLimits()
 			if err != nil {
-				klog.V(4).Infof("Error getting volume limit for plugin %s", volumePlugin.GetPluginName())
+				freqlog.V(4).InfoInfreqf("Error getting volume limit for plugin %s", volumePlugin.GetPluginName())
```
why does this not log the error?

thockin commented on pkg/kubelet/prober/prober.go:
```diff
@@ -170,10 +171,10 @@ func (pb *prober) runProbe(probeType probeType, p *v1.Probe, pod *v1.Pod, status
 			return probe.Unknown, "", err
 		}
 		path := p.HTTPGet.Path
-		klog.V(4).Infof("HTTP-Probe Host: %v://%v, Port: %v, Path: %v", scheme, host, port, path)
```
Why are we rate-limiting V(4) ? Who sees those anyway?

thockin commented on pkg/proxy/healthcheck/healthcheck.go:
```diff
@@ -232,7 +233,7 @@ func (hcs *server) SyncEndpoints(newEndpoints map[types.NamespacedName]int) erro
 	for nsn, count := range newEndpoints {
 		if hcs.services[nsn] == nil {
-			klog.V(3).Infof("Not saving endpoints for unknown healthcheck %q", nsn.String())
+			freqlog.V(3).InfoInfreqf("Not saving endpoints for unknown healthcheck %q", nsn.String())
```
If we are seeing this, I'd say it's a bug

thockin commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -1040,7 +1041,7 @@ func (proxier *Proxier) syncProxyRules() {
 	}
 
-	klog.V(3).Infof("Syncing Policies")
+	freqlog.V(3).InfoInfreqf("Syncing Policies")
```
prefer to just crank up V

thockin commented on pkg/util/async/bounded_frequency_runner.go:
```diff
@@ -218,7 +219,7 @@ func (bfr *BoundedFrequencyRunner) tryRun() {
 		bfr.timer.Stop()
 		bfr.timer.Reset(bfr.maxInterval)
-		klog.V(3).Infof("%s: ran, next possible in %v, periodic in %v", bfr.name, bfr.minInterval, bfr.maxInterval)
+		freqlog.V(3).InfoInfreqf("%s: ran, next possible in %v, periodic in %v", bfr.name, bfr.minInterval, bfr.maxInterval)
```
Also prefer to crank up V

thockin commented on pkg/util/log/logger.go:
```diff
@@ -0,0 +1,109 @@
+	if v {
+		message := fmt.Sprintf(format, args...)
+		if message = getMessageInfrequently(message); message != "" {
+			klog.Infof(message)
```
This will obscure the file:line info of all such log messages. You'd need to use something like `InfoDepth`.  See https://github.com/go-logr/glogr/blob/master/glogr.go

thockin replied:
Also, you already formatted it, so `Infof` and `Errorf` should be `Info` and `Error`

thockin commented on pkg/util/log/logger.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func getMessageInfrequently(message string) string {
+	entry, ok := messages[message]
```
Wouldn't you do a lot better to hash the message, e.g. md5sum or something?

thockin replied:
Also this is a race.  You can't read it without the lock and then write it with the lock.

thockin commented on pkg/util/log/logger.go:
```diff
@@ -0,0 +1,109 @@
+
+	entry.count++
+	now := time.Now()
+	if delta := now.Sub(entry.timestamp); delta >= 5*time.Minute || delta >= time.Duration(entry.countLogged*10)*time.Second {
```
time.Since() is more idiomatic that time.Now().Sub()

thockin replied:
All of these constant values should be named consts

thockin replied:
comments on tricky heuristics, please

thockin commented on pkg/util/log/logger.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+// ErrorInfreqf logs to the ERROR, WARNING, and INFO logs, but identical log messages
+// will not be logged if sooner than an reasonable time window. That time window increases
```
This needs to define better what it is doing

thockin reviewed (commented):
I have to say I am -1 on this approach to logspam.  At least with regards to the messages chosen, I'd much rather fix the verbosity levels of them or fix the problems that cause them to get logged.

I reviewed this as if we would be doing it, but I am not sure we should.
